### PR TITLE
Make TaskBatcher Less Lock-Heavy

### PR DESCRIPTION
**Pull request by original-brownbear**
In many shards benchmarks we see a lot of contention when submitting tasks.
This is obvious when working with lots of large task batches and doing long iterations.

We don't need to lock in `runIfNotProcessed` past removing the task set for a key
and can be a little more efficient when it comes to creating the new tasks set in
`submitTasks` as well.
Also, we don't need to use a fully locking map as we often have operations for different
batching keys interleaved so moving to CHM as well.

This change is particularly relevant for stability because we often submit tasks from
network threads directly where grinding through a e.g. bunch of shard state updates and
having to lock on the map over and over while e.g. a huge batch of index create or so
was iterated over in `runIfNotProcessed` caused very visible latency.

relates #77466

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Also relates https://github.com/elastic/elasticsearch/issues/81626. I'm unlikely to get to this today, bear with me.

File: server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java [hunk 1]
```diff
@@ -61,11 +62,10 @@ public void submitTasks(List<? extends BatchedTask> tasks, @Nullable TimeValue t
...
-            LinkedHashSet<BatchedTask> existingTasks = tasksPerBatchingKey.computeIfAbsent(
-                firstTask.batchingKey,
-                k -> new LinkedHashSet<>(tasks.size())
-            );
+        tasksPerBatchingKey.compute(firstTask.batchingKey, (k, existingTasks) -> {
+            if (existingTasks == null) {
+                return new LinkedHashSet<>(tasks);
+            }
             for (BatchedTask existing : existingTasks) {
                 // check that there won't be two tasks with the same identity for the same batching key
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java:70]: FWIW I don't think I've ever seen this check fail, maybe we could make it an assertion?

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java:70]: I'm all for it. I think we can only get here via a bug and I've also never seen this happen outside of bugs during experimenting with stuff. I'll make it an assertion :)

**Comment by original-brownbear** [comment]
> I'm unlikely to get to this today, bear with me.

No rush :) + Made the dup. check an assertion now.

File: server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java [hunk 2]
```diff
@@ -51,42 +52,47 @@ public void submitTasks(List<? extends BatchedTask> tasks, @Nullable TimeValue t
         assert tasks.stream().allMatch(t -> t.batchingKey == firstTask.batchingKey)
             : "tasks submitted in a batch should share the same batching key: " + tasks;
         // convert to an identity map to check for dups based on task identity
+
+        tasksPerBatchingKey.compute(firstTask.batchingKey, (k, existingTasks) -> {
+            assert assertNoDuplicateTasks(tasks, existingTasks);
+            if (existingTasks == null) {
+                return new LinkedHashSet<>(tasks);
+            }
+            existingTasks.addAll(tasks);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java:61]: I am not too fond of this pattern. I *think* it works due to how CHM synchronizes on the head of the bucket in both compute and remove. But I think there is no guarantee from CHM to do either. It seems unlikely that compute could run the remapping function on the same key in parallel (since that would break the call only once guarantee), but I am less certain that a future evolution of CHM could not remove the synchronized in remove (do not see a way to do it though).

In short, I think we rely too much on the internals of CHM here in a very central place. Could we perhaps go immutable here and return a list of the two lists - and flatten it when extracted/removed?

**Review by henningandersen — COMMENTED** [review]
I wonder if we can use an immutable (or alternatively concurrent) structure in the map?

File: server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java:61]: I think immutable either gets quite tricky (if we keep a list of lists, the timeout task removal gets awkward) or slow if we copy over and over. I'd rather go synchronized I think, it shouldn't be too much overhead here given how we will never have contention on the collection.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java:62]: Though tbh. I'm not sure about the need for this. CHM docs state:

> This class obeys the same functional specification as {@link java.util.Hashtable}
 
which means that remove will always be synchronized with the compute operation I'd assume?

**Comment by original-brownbear** [comment]
Made the collection synchronized now 56ba16c  let me know what you think :)

File: server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java [hunk 3]
```diff
@@ -127,17 +132,15 @@ void runIfNotProcessed(BatchedTask updateTask) {
...
-                        if (task.processed.getAndSet(true) == false) {
-                            logger.trace("will process {}", task);
-                            toExecute.add(task);
-                            processTasksBySource.computeIfAbsent(task.source, s -> new ArrayList<>()).add(task);
-                        } else {
-                            logger.trace("skipping {}, already processed", task);
-                        }
+            final Set<BatchedTask> pending = tasksPerBatchingKey.remove(updateTask.batchingKey);
+            if (pending != null) {
+                for (BatchedTask task : pending) {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java:137]: I think we would want to synchronize explicitly on the `pending` set instead of relying on `Collections.synchronizedSet`, iterating a synchronized set like this is not really safe.

File: server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java (hunk 2 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java:62]: It says right after that:

```
This class is fully
 * interoperable with {@code Hashtable} in programs that rely on its
 * thread safety but not on its synchronization details.
```

It is also just nice to have this code "obviously correct", since it is so important here.

**Review by henningandersen — COMMENTED** [review]
Thanks, I think synchronized is fine, though we should explicitly synchronize instead.

File: server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java]: Right fair point :) Added locking to this loop.

**Review by henningandersen — APPROVED** [review]
LGTM.

**Comment by original-brownbear** [comment]
Thanks Henning + David!
